javascript/2306-naming-a-company.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * @param {string[]} ideas
+ * @return {number}
+ */
+
+// Time complexity: O(n^2 * m)
+// Space complexity: O(n * m)
+
+var distinctNames = function (ideas) {
+    const wordMap = new Map();
+    let count = 0;
+
+    for (const word of ideas) {
+        const ch = word.charAt(0);
+        const substr = word.slice(1);
+        const set = wordMap.get(ch) ?? new Set();
+        set.add(substr);
+        wordMap.set(ch, set);
+    }
+
+    for (const [ch1, set1] of wordMap) {
+        for (const [ch2, set2] of wordMap) {
+            if (ch1 === ch2) {
+                continue;
+            }
+
+            const intersect = getIntersection(set1, set2);
+            const distinct1 = set1.size - intersect;
+            const distinct2 = set2.size - intersect;
+            count += distinct1 * distinct2;
+        }
+    }
+
+    return count;
+};
+
+function getIntersection(set1, set2) {
+    let intersect = 0;
+    for (const word of set1) {
```

Comment:
I suggested avoid functional looping, but in this case you can take advantage of filter when doing set intersections

```suggestion
comet getIntersection = (set1, set2) => [ ...set1 ]
    .filter((key) => set2.has(key))
    .length;
```

Reply:
here as well updated the code.